### ISSUE TITLE
map_index_template not compute when task marked success manually

### ISSUE THREAD

**Issue by raphaelauv** [apache/airflow#39092]
### Apache Airflow version

2.9.0

### What happened?

If i mark a task manually at success while it's running only already succeed mapped task have a `map_index_template`

![Screenshot from 2024-04-17 18-13-16](https://github.com/apache/airflow/assets/10202690/48fa5199-2401-434d-899c-698a78ae4ce9)

### What you think should happen instead?

`map_index_template`

should be compute in all cases

### How to reproduce

```python
from airflow.operators.python import PythonOperator

from pendulum import today
from airflow import DAG

dag_1 = DAG(
    dag_id="dag_1",
    schedule_interval=None,
    max_active_tasks=2,
    start_date=today("UTC").add(days=-1)
)

with dag_1:
    def generate():
        return [{'date': '2024-01-01'}, {'date': '2024-01-02'}, {'date': '2024-01-03'}, {'date': '2024-01-04'},
                {'date': '2024-01-05'}, {'date': '2024-01-06'}, {'date': '2024-01-07'}]


    def toto(date, arg_2=None):
        import time
        print(date)
        time.sleep(10)


    a = PythonOperator(
        task_id="a",
        python_callable=generate)
    b = PythonOperator.partial(
        task_id="b",
        map_index_template="{{ task.op_kwargs['date'] }}",
        python_callable=toto).expand(op_kwargs=a.output)

    a >> b

```

trigger a run  , wait for a mapped task to succes , then mark state of the task `b` at success

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by RNHTTR** [comment]
This _might_ be fixed by #38902 . It might require adding some logic to render mapped tasks in [set_task_instance_state](https://github.com/apache/airflow/blob/066708352e6a6a06f213b65324e982f582019b8e/airflow/models/dag.py#L2078), which I believe is called when manually setting a task instance's state.

**Comment by karenbraganz** [comment]
I would like to be assigned this issue.

**Comment by raphaelauv** [comment]
@karenbraganz are you still working on this issue ? thanks

**Comment by karenbraganz** [comment]
I created a draft PR but was informed that rendering the Jinja template on the webserver is not viable. @uranusjr could you please clarify why this is the case? I accidentally closed the PR before I could get a complete explanation from you. Here is the PR for reference: https://github.com/apache/airflow/pull/39505

**Comment by raphaelauv** [comment]
`accidentally closed the PR` you can't re-open it ?

**Comment by karenbraganz** [comment]
I did not re-open it because it looks like the functionality is not viable. The only thing left was for @uranusjr to provide an explanation of why they think it's not viable, which is why I have tagged them here for an explanation.

**Comment by raphaelauv** [comment]
could you please tell @RNHTTR  if it's okay to unassign you

**Comment by uranusjr** [comment]
Rendering Jinja templates in the webserver means injecting custom user logic into protected Airflow services, which is not acceptable. Rendering must happen in the worker. A fix along the line mentioned in Ryan’s comment above is more likely correct, although I am not sure (and never checked) where exactly the rendering should happen.

**Comment by karenbraganz** [comment]
I got the error referenced in my PR after adding my rendering logic to set_task_instance_state() in models/dag.py. I just tried adding the logic to _set_state() (from models/taskinstance.py) which is called by set_task_instance_state() in www/views.py and am getting the same error. I'm not sure how to get the worker to execute the code at the moment, so I will unassign myself.

**Comment by uranusjr** [comment]
I wonder if we should simply provide a way for the user to set the map index to whatever they want manually in the web UI. This works around the lack of worker situation, and you can even _change_ the map index after the fact if you so wish.

This does not completely solve the problem (setting the task state should still trigger a worker in some cases), but should be good enough for many use cases.

**Comment by raphaelauv** [comment]
would it be possible or "correct" to let the webserver show the value of map_index of task when it's a simple templating like , cause the information is already available 

```map_index_template="{{ task.op_kwargs['date'] }}",```

**Comment by uranusjr** [comment]
But how do you detect this? A sufficiently knowledgable Airflow user can reason with this, but it’d be quite difficult and error-prone to implement all the logic.

**Comment by raphaelauv** [comment]
then a try catch of rendering ?

 like it's already the case for the args of the operator 

![Image](https://github.com/user-attachments/assets/2047e734-9111-4437-991a-b067852a960a)

**Comment by uranusjr** [comment]
Wait, you mean to display the template string literally, without any rendering? I don’t think that’s an overall improvement tbh since now you can’t even tell the task instances apart from each other. Maybe it’d be useful in some context, but I can’t see it replacing the simple integer.

**Comment by raphaelauv** [comment]
No, I have in mind , the webserver could try to render and if it fails cause it's a complex case then just show a warning and the default value

**Comment by uranusjr** [comment]
I don’t think it’s a good idea to try to render the web server. Jinja2 templates can contain arbitrarily complex untrusted code, and should not be run in the web server process for the same reason as dag code.

**Comment by kaxil** [comment]
Closed by https://github.com/apache/airflow/pull/57208
